### PR TITLE
Add automatic addon check(Solves #9)

### PR DESCRIPTION
**Pull request by LordGameleo**
Add automatic addon check(Solves Issue #9 )

### CODE REVIEW

**Comment by LordGameleo** [comment]
@Razzeee  addon is supports only python3 it doesn't support python2.7

**Comment by LordGameleo** [comment]
we can remove python2.7 from travis

**Comment by Razzeee** [comment]
If that's the case, then it's a bug that needs fixing. As it should in the best case run on a python subset that works on both 2 and 3

**Comment by LordGameleo** [comment]
@Razzeee So what should be done here? We can create an issue in addon check to support both.

File: .travis.yml [hunk 1]
```diff
@@ -10,9 +10,25 @@ matrix:
...
+    cd $HOME
+    git clone https://github.com/xbmc/addon-check
+    cd $TRAVIS_BUILD_DIR/
+
 # command to install dependencies
-install: 
-  - pip install pytest mock
+install:
+  - |
+    pip install $HOME/addon-check
```
Comment (Rechi) [inline .travis.yml]:
```suggestion
    pip install kodi-addon-checker
```

File: .travis.yml [hunk 2]
```diff
@@ -10,9 +10,25 @@ matrix:
...
+    pip install pytest mock
+
+before_script:
+  - |
+    rm -rf LICENSES/
+    rm -rf $HOME/addon-check
+    cd $TRAVIS_BUILD_DIR/
+    kodi-addon-checker $TRAVIS_BUILD_DIR/notrobro-detector/ --branch=leia
```
Comment (Rechi) [inline .travis.yml]: This is not a command expected in `before_script` block, it should be in `script`. Also it checks the wrong directory as `service.notrobro` contains the addon, not `notrobro-detector`.

**Comment by Rechi** [comment]
> We can create an issue in addon check to support both.

That won't happen. `addon-check` will never support Python 2. Just run it only with a Python 3 version.
Also there is no benefit in running it with every Python3 variant as the ouput will always be the same.

**Comment by LordGameleo** [comment]
@Rechi  Should we remove Python2.7 from travis?

**Comment by Rechi** [comment]
No, the addon itself still has to be compatible with Python 2.7 for Kodi v18 (Leia).

File: .travis.yml [hunk 3]
```diff
@@ -1,7 +1,6 @@
 language: python
 matrix:
   include:
-    - python: "2.7"
```
Comment (Rechi) [inline .travis.yml]: No, Python 2.7 can't  be removed. Only kodi-addon-checker should not be run in this job.

**Comment by LordGameleo** [comment]
@Rechi  kodi-addon-checker is failing with errors.

File: .travis.yml [hunk 4]
```diff
@@ -10,9 +9,17 @@ matrix:
...
     - python: "nightly"
-  
+
 # command to install dependencies
-install: 
-  - pip install pytest mock
+install:
+  - |
+    pip install kodi-addon-checker
+    pip install pytest mock
```
Comment (Rechi) [inline .travis.yml]: No need to make this a single 'command'.
```suggestion
install:
    - pip install kodi-addon-checker
    - pip install pytest mock
```

**Comment by LordGameleo** [comment]
> No, the addon itself still has to be compatible with Python 2.7 for Kodi v18 (Leia).

I didn't get you. If addon is not compatible with Python 2.7 then why have travis check for that build

File: .travis.yml [hunk 5]
```diff
@@ -10,9 +9,17 @@ matrix:
...
+
+before_script: rm -rf LICENSES/
```
Comment (Rechi) [inline .travis.yml]: This directory doesn't exists.

File: .travis.yml [hunk 6]
```diff
@@ -10,9 +9,17 @@ matrix:
...
+
 # command to run tests
-script: py.test -v
+script:
+  - |
+    py.test -v
+    kodi-addon-checker $TRAVIS_BUILD_DIR/service.notrobro/ --branch=leia
```
Comment (Rechi) [inline .travis.yml]: No need to make this a single 'command'.
```suggestion
script:
  - py.test -v
  - kodi-addon-checker --branch=leia -- service.notrobro/
```

**Comment by Rechi** [comment]
The addon is compatible with Python 2.7. The **addon-checker tool** isn't Python 2.7 compatible. Because of that do **not** run this tool with Python 2.7.

File: .travis.yml [hunk 7]
```diff
@@ -10,9 +9,21 @@ matrix:
...
   allow_failures:
     - python: "nightly"
-  
+
 # command to install dependencies
-install: 
-  - pip install pytest mock
+install:
+  - |
+    pip install kodi-addon-checker
```
Comment (Rechi) [inline .travis.yml]: Now there are two install steps.

File: .travis.yml [hunk 8]
```diff
@@ -10,9 +9,21 @@ matrix:
...
+    - pip install kodi-addon-checker
+    - pip install pytest mock
+
+before_script: rm -rf LICENSES/
+
 # command to run tests
-script: py.test -v
+script:
+  - |
+    py.test -v
```
Comment (Rechi) [inline .travis.yml]: Now there are two script steps.

File: .travis.yml [hunk 9]
```diff
@@ -10,9 +10,13 @@ matrix:
       sudo: required
   allow_failures:
     - python: "nightly"
-  
+
 # command to install dependencies
-install: 
-  - pip install pytest mock
+install:
+    - pip install kodi-addon-checker
```
Comment (Rechi) [inline .travis.yml]:
```suggestion
    - if [[ $TRAVIS_PYTHON_VERSION != '2.7' ]]; then pip install kodi-addon-checker; fi
```

File: .travis.yml [hunk 10]
```diff
@@ -10,9 +10,13 @@ matrix:
...
+    - pip install pytest mock
+
 # command to run tests
-script: py.test -v
+script:
+  - py.test -v
+  - kodi-addon-checker --branch=leia -- service.notrobro/
```
Comment (Rechi) [inline .travis.yml]:
```suggestion
  - if [[ $TRAVIS_PYTHON_VERSION != '2.7' ]]; then kodi-addon-checker --branch=leia -- service.notrobro/; fi
```

**Comment by LordGameleo** [comment]
@Rechi it is failing because service.notrobro is not a valid addon. It doesn't contain any fanart, icon and schema is also failing

**Comment by Rechi** [comment]
Can you fix those errors?

**Comment by LordGameleo** [comment]
@Rechi I will do it :+1:

**Comment by LordGameleo** [comment]
@Rechi Should I use any fanart and icon?

**Comment by Rechi** [comment]
No, just remove the definitions that the addon has an icon and fanart.

**Comment by LordGameleo** [comment]
@Rechi how can below error be corrected..... 
```
failed validating {'point': 'xbmc.service', 'library': 'main.py', 'start': 'login'} with XsdAttributeGroup(['point', 'id', 'name', 'library']):

Reason: 'start' attribute not allowed for element.

Schema:

  <xs:complexType xmlns:xs="http://www.w3.org/2001/XMLSchema">
    <xs:attribute name="point" type="xs:string" use="required" />
    <xs:attribute name="id" type="simpleIdentifier" />
    <xs:attribute name="name" type="xs:string" />
    <xs:attribute name="library" type="xs:string" use="required" />
  </xs:complexType>

Instance:

  <extension library="main.py" point="xbmc.service" start="login" />

Path: /extension

failed validating '' with XsdPatternFacets(['\\s*([^\\s]\\s*)+']):

Reason: value doesn't match any pattern of ['\\s*([^\\s]\\s*)+'].

Schema:

  <xs:pattern xmlns:xs="http://www.w3.org/2001/XMLSchema" value="\s*([^\s]\s*)+" />

Instance:

  <source />

Path: /extension/source

failed validating '' with XsdPatternFacets(['\\s*([^\\s]\\s*)+']):

Reason: value doesn't match any pattern of ['\\s*([^\\s]\\s*)+'].

Schema:

  <xs:pattern xmlns:xs="http://www.w3.org/2001/XMLSchema" value="\s*([^\s]\s*)+" />

Instance:

  <news />

Path: /extension/news

failed validating '' with XsdPatternFacets(['\\s*([^\\s]\\s*)+']):

Reason: value doesn't match any pattern of ['\\s*([^\\s]\\s*)+'].

Schema:

  <xs:pattern xmlns:xs="http://www.w3.org/2001/XMLSchema" value="\s*([^\s]\s*)+" />

Instance:

  <news />

Path: /extension/news

failed validating {} with XsdAttributeGroup(['lang']):

Reason: missing required attributes: {'lang'}

Schema:

  <xs:extension xmlns:xs="http://www.w3.org/2001/XMLSchema" base="nonEmptyStringCapped">
    <xs:attribute name="lang" type="langIdentifier" use="required" />
  </xs:extension>

Instance:

  <disclaimer />

Path: /extension/disclaimer

ERROR: Schema validation failed for the following points: xbmc.servicexbmc.addon.metadata
```

**Comment by Rechi** [comment]
- start attribute isn't allowed for leia
- source, news and disclaimer must have a non empty value, if they are specified
- disclaimer would also need the `lang` attribute

**Comment by LordGameleo** [comment]
- Should I remove start attribute
- Should I add random values for source, news and disclaimer or should I copy it from another addon?

**Comment by Rechi** [comment]
> Should I remove start attribute

As it isn't allowed anymore, yes. Or how would you fix it otherwise

> Should I add random values for source, news and disclaimer or should I copy it from another addon?

How are random or wrong values more helpful then none?

**Comment by LordGameleo** [comment]
> How are random or wrong values more helpful then none?

It will be just a temporary solution to pass the test cases

**Comment by Rechi** [comment]
Then you can also remove the tags, which doesn't give the impression that those values are already filled out and also fixes the test cases.

**Comment by LordGameleo** [comment]
I am getting an error while checking it using `kodi-addon-checker service.notrobro/ --branch=leia` 

```
Traceback (most recent call last):
  File "/home/glord/Desktop/Kodi/notro/lib/python3.6/site-packages/urllib3/connectionpool.py", line 384, in _make_request
    six.raise_from(e, None)
  File "<string>", line 2, in raise_from
  File "/home/glord/Desktop/Kodi/notro/lib/python3.6/site-packages/urllib3/connectionpool.py", line 380, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib/python3.6/http/client.py", line 1331, in getresponse
    response.begin()
  File "/usr/lib/python3.6/http/client.py", line 297, in begin
    version, status, reason = self._read_status()
  File "/usr/lib/python3.6/http/client.py", line 258, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/lib/python3.6/socket.py", line 586, in readinto
    return self._sock.recv_into(b)
socket.timeout: timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/glord/Desktop/Kodi/notro/bin/kodi-addon-checker", line 10, in <module>
    sys.exit(main())
  File "/home/glord/Desktop/Kodi/notro/lib/python3.6/site-packages/kodi_addon_checker/__main__.py", line 85, in main
    all_repo_addons = check_addon.all_repo_addons()
  File "/home/glord/Desktop/Kodi/notro/lib/python3.6/site-packages/kodi_addon_checker/check_addon.py", line 119, in all_repo_addons
    repo_addons[branch] = Repository(branch, branch_url)
  File "/home/glord/Desktop/Kodi/notro/lib/python3.6/site-packages/kodi_addon_checker/addons/Repository.py", line 22, in __init__
    gz_file = requests.get(path, timeout=(10, 10)).content
  File "/home/glord/Desktop/Kodi/notro/lib/python3.6/site-packages/requests/api.py", line 75, in get
    return request('get', url, params=params, **kwargs)
  File "/home/glord/Desktop/Kodi/notro/lib/python3.6/site-packages/requests/api.py", line 60, in request
    return session.request(method=method, url=url, **kwargs)
  File "/home/glord/Desktop/Kodi/notro/lib/python3.6/site-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/glord/Desktop/Kodi/notro/lib/python3.6/site-packages/requests/sessions.py", line 668, in send
    history = [resp for resp in gen] if allow_redirects else []
  File "/home/glord/Desktop/Kodi/notro/lib/python3.6/site-packages/requests/sessions.py", line 668, in <listcomp>
    history = [resp for resp in gen] if allow_redirects else []
  File "/home/glord/Desktop/Kodi/notro/lib/python3.6/site-packages/requests/sessions.py", line 247, in resolve_redirects
    **adapter_kwargs
  File "/home/glord/Desktop/Kodi/notro/lib/python3.6/site-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/home/glord/Desktop/Kodi/notro/lib/python3.6/site-packages/requests/adapters.py", line 529, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPConnectionPool(host='ftp.snt.utwente.nl', port=80): Read timed out. (read timeout=10)
```

**Comment by Rechi** [comment]
Run the command again.

**Comment by LordGameleo** [comment]
Did that multiple times. Same error

**Comment by LordGameleo** [comment]
I pushed my code, now travis also showing same error.

File: service.notrobro/addon.xml [hunk 11]
```diff
@@ -19,8 +19,6 @@
         <news></news>
```
Comment (Rechi) [inline service.notrobro/addon.xml]:
```suggestion
```

File: service.notrobro/addon.xml [hunk 12]
```diff
@@ -19,8 +19,6 @@
         <news></news>
         <disclaimer></disclaimer>
```
Comment (Rechi) [inline service.notrobro/addon.xml]:
```suggestion
```

File: service.notrobro/addon.xml [hunk 13]
```diff
@@ -19,8 +19,6 @@
         <news></news>
         <disclaimer></disclaimer>
         <assets>
```
Comment (Rechi) [inline service.notrobro/addon.xml]: Also remove the assets tag, if it has no child elements.
```suggestion
```

File: service.notrobro/addon.xml [hunk 14]
```diff
@@ -19,8 +19,6 @@
         <news></news>
         <disclaimer></disclaimer>
         <assets>
-            <icon>resources/icon.png</icon>
-            <fanart>resources/fanart.jpg</fanart>
         </assets>
```
Comment (Rechi) [inline service.notrobro/addon.xml]:
```suggestion
```

**Comment by LordGameleo** [comment]
@Rechi  I removed them. Now the build is success.

**Comment by LordGameleo** [comment]
Hey @Razzeee  and @Rechi can I know your timezone and when you are more active. I will try to sync my timings with you.

**Comment by hudokkow** [comment]
UTC/GMT +1

File: .travis.yml [hunk 15]
```diff
@@ -10,9 +10,13 @@ matrix:
...
-  - pip install pytest mock
+install:
+    - pip install pytest mock
+    - if [[ $TRAVIS_PYTHON_VERSION != '2.7' ]]; then pip install kodi-addon-checker; fi
+
 # command to run tests
-script: py.test -v
+script:
+  - py.test -v
+  - if [[ $TRAVIS_PYTHON_VERSION != '2.7' ]]; then kodi-addon-checker --branch=leia  -- service.notrobro/; fi
```
Comment (Rechi) [inline .travis.yml]: I would swap those two lines, to get rid of `Found non whitelisted file ending in filename *.pyc` warnings.

File: service.notrobro/addon.xml [hunk 16]
```diff
@@ -5,22 +5,13 @@
...
     <extension point="xbmc.addon.metadata">
         <summary lang="en_GB">Skip intros and outros</summary>
         <description lang="en_GB"></description>
-        <language></language>
         <platform>all</platform>
         <license>GPL-3.0</license>
-        <forum></forum>
         <website>www.kodi.tv</website>
         <email>mohit15060@iiitd.ac.in</email>
-        <source></source>
```
Comment (Rechi) [inline service.notrobro/addon.xml]: The source location is obvious.
```suggestion
        <source>https://github.com/xbmc/notrobro</source>
```

**Comment by LordGameleo** [comment]
@hudokkow is it same for everyone?

File: service.notrobro/addon.xml (hunk 16 above)
Comment (LordGameleo) [reply service.notrobro/addon.xml]: @Rechi should I remove it?

Comment (Rechi) [reply service.notrobro/addon.xml]: You should readd source tag with the correct value.

Comment (LordGameleo) [reply service.notrobro/addon.xml]: it should be...... https://github.com/xbmc/notrobro//service.notrobro ?

Comment (Rechi) [reply service.notrobro/addon.xml]: That URL doesn't exist. It should be https://github.com/xbmc/notrobro. Everyone should be able to find the service.notrobro folder there.

**Comment by hudokkow** [comment]
For @Razzeee and @Rechi yes. For me it's UTC. Team members are spread around the world but many (most?) are in Europe.

**Comment by LordGameleo** [comment]
@hudokkow 
Is there any other medium to connect with kodi team members except fourm?

**Comment by hudokkow** [comment]
Join #kodi-gsoc on https://webchat.freenode.net/ or connect with your favorite IRC client.

File: .travis.yml [hunk 17]
```diff
@@ -10,9 +10,14 @@ matrix:
...
-install: 
-  - pip install pytest mock
+install:
+    - pip install pytest mock
+    - if [[ $TRAVIS_PYTHON_VERSION != '2.7' ]]; then pip install kodi-addon-checker; fi
+
 # command to run tests
-script: py.test -v
+script:
+  - if [[ $TRAVIS_PYTHON_VERSION != '2.7' ]]; then kodi-addon-checker --branch=leia  -- service.notrobro/; fi
```
Comment (Razzeee) [inline .travis.yml]: can we separate unit test steps from integration tests into different stages.

that way we should be able to overwrite the matrix values like this https://docs.travis-ci.com/user/build-stages/matrix-expansion/

File: service.notrobro/addon.xml (hunk 16 above)
Comment (LordGameleo) [reply service.notrobro/addon.xml]: @Rechi  https://github.com/xbmc/notrobro/service.notrobro should be the url right?

**Comment by LordGameleo** [comment]
@Razzeee I don't know why but travis is not building.

**Comment by Razzeee** [comment]
might be an invalid config? you should be able test it if you activate travis for your account.

**Comment by LordGameleo** [comment]
Okay

File: .travis.yml (hunk 17 above)
Comment (LordGameleo) [reply .travis.yml]: Do you want something like this?? @Razzeee

File: .travis.yml [hunk 18]
```diff
@@ -1,18 +1,83 @@
...
-# command to run tests
-script: py.test -v
+      python:
+        - "2.7"
+      install:
+        - pip install kodi-addon-checker
+        - pip install pytest mock
+
+      script:
+        - echo Deploy
```
Comment (Rechi) [inline .travis.yml]: `py.test -v` doesn't do deploying.

File: .travis.yml [hunk 19]
```diff
@@ -1,18 +1,83 @@
...
+        - py.test -v
+      allow_failures:
+        - python: "nightly"
```
Comment (Rechi) [inline .travis.yml]: Python nightly version isn't used.

File: .travis.yml [hunk 20]
```diff
@@ -1,18 +1,83 @@
...
+
+    - stage: Python_3
+      env: Py3.4
+      dist: xenial
```
Comment (Rechi) [inline .travis.yml]: move this at the top (like `language: python`) so it doesn't have to be repeated everywhere

File: .travis.yml (hunk 19 above)
Comment (LordGameleo) [reply .travis.yml]: I removed it.

File: .travis.yml (hunk 18 above)
Comment (LordGameleo) [reply .travis.yml]: @Rechi I am sorry that piece of code was there by mistake.

**Comment by LordGameleo** [comment]
@Razzeee @Rechi Please Review.

File: service.notrobro/addon.xml [hunk 21]
```diff
@@ -5,22 +5,14 @@
...
-        <news></news>
-        <disclaimer></disclaimer>
-        <assets>
```
Comment (Razzeee) [inline service.notrobro/addon.xml]: These should not be removed

**Comment by Razzeee** [comment]
Why are there binaries included in the PR now? Probably on accident?

File: .travis.yml [hunk 22]
```diff
@@ -1,18 +1,37 @@
...
+sudo: required
+install:
   - pip install pytest mock
-# command to run tests
-script: py.test -v
+
+jobs:
+  include:
+    - stage: Tests
+      python: "2.7"
```
Comment (Razzeee) [inline .travis.yml]: Did you try to add multiple pythons here? Would be surprise if that didn't work.

I would expect this but haven't looked at docs:
```
python:
     - "2.7"
     - "3.4"
     - "..."
```

File: service.notrobro/addon.xml (hunk 21 above)
Comment (LordGameleo) [reply service.notrobro/addon.xml]: These are just empty values. Why do we need them

File: .travis.yml (hunk 22 above)
Comment (LordGameleo) [reply .travis.yml]: @Razzeee it is not giving the desired results.

**Comment by LordGameleo** [comment]
@Rechi @Razzeee Please Review.

**Comment by LordGameleo** [comment]
> Why are there binaries included in the PR now? Probably on accident?

Yeah they were there by mistake. I removed them in latest commit.

File: .travis.yml (hunk 22 above)
Comment (Rechi) [reply .travis.yml]:
```
language: python
dist: xenial
sudo: required

python:
  - 2.7
  - 3.4
  - 3.5
  - 3.6
  - 3.7

# command to install dependencies
install:
  - pip install pytest mock

# command to run tests
script: py.test -v

jobs:
  include:
    - stage: Kodi-Addon-Check
      python: 3.6

      # command to install dependencies
      install:
        - pip install kodi-addon-checker
      # command to run tests
      script:
        - kodi-addon-checker --branch=leia  -- service.notrobro/;
```

File: service.notrobro/addon.xml (hunk 21 above)
Comment (Razzeee) [reply service.notrobro/addon.xml]: assets are not empty, at least that should be kept. I think it's even mandatory for leia now.

Comment (LordGameleo) [reply service.notrobro/addon.xml]: I had to remove content of assets i.e. fanart and icon as it was casuing it to fail.

Comment (Razzeee) [reply service.notrobro/addon.xml]: If that's reproduce-able it might be a bug in the checker.

Comment (Rechi) [reply service.notrobro/addon.xml]: Adding an empty assets tag has a different meaning then no assets tag.
- empty: no icon, fanart and screenshot exists
- not listed: try icon and fanart fallback locations

**Comment by Rechi** [comment]
@LordGameleo please squash the commits. Then it's okay for me.

**Comment by LordGameleo** [comment]
@Rechi All done. Please check.

**Comment by Rechi** [comment]
binary files are back

**Comment by LordGameleo** [comment]
@Rechi Removed binary files now.

**Comment by LordGameleo** [comment]
@Razzeee changes have been approved by Rechi.

**Pull request merged**